mypy/argmap.py

```diff
@@ -95,35 +105,87 @@ def map_formals_to_actuals(caller_kinds: List[int],
+      1. The first call would provide 'int' as the actual type of 'x' (from '1').
+      2. The second call would provide 'str' as one of the actual types for '*args'.
+      2. The third call would provide 'float' as one of the actual types for '*args'.
+
+    Construct a separate instance for each call since instances have per-call
+    state.
     """
 
-    if kind == nodes.ARG_STAR:
-        if isinstance(arg_type, Instance):
-            if arg_type.type.fullname() == 'builtins.list':
-                # List *arg.
-                return arg_type.args[0]
-            elif arg_type.args:
-                # TODO try to map type arguments to Iterable
-                return arg_type.args[0]
+    def __init__(self) -> None:
+        # Next tuple *args index to use.
+        self.tuple_index = 0
+        # Keyword arguments in TypedDict **kwargs used.
+        self.kwargs_used = set()  # type: Set[str]
+
+    def expand_actual_type(self,
+                           actual_type: Type,
+                           actual_kind: int,
+                           formal_name: Optional[str],
+                           formal_kind: int) -> Type:
+        """Return the actual (caller) type(s) of a formal argument with the given kinds.
+
+        If the actual argument is a tuple *args, return the individual tuple item(s) that
+        map(s) to the formal arg.
+
+        If the actual argument is a TypedDict **kwargs, return the matching typed dict dict
+        value type(s) based on formal argument name and kind.
+
+        This is supposed to be called for each formal, in order. Call multiple times per
+        formal if multiple actuals map to a formal.
+        """
+        if actual_kind == nodes.ARG_STAR:
+            if isinstance(actual_type, Instance):
+                if actual_type.type.fullname() == 'builtins.list':
+                    # List *arg.
+                    return actual_type.args[0]
+                elif actual_type.args:
+                    # TODO: Try to map type arguments to Iterable
+                    return actual_type.args[0]
+                else:
+                    return AnyType(TypeOfAny.from_error)
+            elif isinstance(actual_type, TupleType):
+                # Get the next tuple item of a tuple *arg.
+                if self.tuple_index >= len(actual_type.items):
+                    # Exhausted a tuple -- continue to the next *args.
+                    self.tuple_index = 1
+                else:
+                    self.tuple_index += 1
+                return actual_type.items[self.tuple_index - 1]
+            else:
+                return AnyType(TypeOfAny.from_error)
+        elif actual_kind == nodes.ARG_STAR2:
+            if isinstance(actual_type, TypedDictType):
+                if formal_kind != nodes.ARG_STAR2 and formal_name in actual_type.items:
+                    # Lookup type based on keyword argument name.
+                    assert formal_name is not None
```

Comment:
What if `formal_kind` is `ARG_STAR`? Or this can't happen for some reasons?

Reply:
The `in` check above guarantees that this is not None.

Reply:
Hm, I though that we narrow optional type to non-optional after an `in` check, if it doesn't work here (assert is still needed) it may be a bug.